### PR TITLE
ELPP-3356 Published articles in RSS feed

### PR DESCRIPTION
**Pull request by giorgiosironi**
Depends on a few other PRs to become green, see ticket.

### CODE REVIEW

File: spectrum/checks.py [hunk 1]
```diff
@@ -669,6 +669,35 @@ def recommend(self, manuscript='', subject='', keywords='', abstract=''):
...
+        )
+
+    def _is_present(self, url, id):
+        response = requests.get(url)
+        LOGGER.debug("Loaded %s (%s)", url, response.status_code)
+        if response.status_code > 299:
+            raise UnrecoverableError(response)
+        soup = BeautifulSoup(response.content, "lxml-xml")
+        target_guid = "https://dx.doi.org/10.7554/eLife.%s" % id
+        guids = {item.guid.string:item for item in soup.rss.channel.find_all("item")}
```
Comment (giorgiosironi) [inline spectrum/checks.py]: I'm a bit concerned by the length of the feed, which currently shows 8 articles in end2end. During the test suite 15-20 articles are published, sometimes in parallel, so with unlucky timing the one published in the new test could slip down and out of the list before it is seen here. Is there a pagination or a configuration we could set up observer with to make sure the list is long enough?

Comment (lsh-0) [reply spectrum/checks.py]: the rss feeds are just reports using rss to serialise their results. all reports support pagination and changing the number of results returned. The reports are not part of the api raml but I'm using the same pagination parameters. Example: 
https://observer.elifesciences.org/report/latest-articles?per-page=1&page=20

most reports have multiple supported serialisations as well as a default serialisation. the default for latest and upcoming articles is rss, but they also support csv: 
https://observer.elifesciences.org/report/latest-articles?page=20&format=json

File: spectrum/checks.py [hunk 2]
```diff
@@ -669,6 +669,35 @@ def recommend(self, manuscript='', subject='', keywords='', abstract=''):
...
+            id, url
+        )
+
+    def _is_present(self, url, id):
+        response = requests.get(url)
+        LOGGER.debug("Loaded %s (%s)", url, response.status_code)
+        if response.status_code > 299:
+            raise UnrecoverableError(response)
+        soup = BeautifulSoup(response.content, "lxml-xml")
+        target_guid = "https://dx.doi.org/10.7554/eLife.%s" % id
```
Comment (lsh-0) [inline spectrum/checks.py]: zero-padding shouldn't be necessary for the id (msid) here, right?

Comment (giorgiosironi) [reply spectrum/checks.py]: Don't think so, because they are all very long.

**Comment by giorgiosironi** [comment]
We still have a problem with the field size:
```
ERROR - unhandled exception attempting to download and regenerate article 3231380270525815893
Traceback (most recent call last):
  File "/srv/observer/venv/lib/python3.5/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/srv/observer/venv/lib/python3.5/site-packages/newrelic-2.78.0.57/newrelic/hooks/database_dbapi2.py", line 22, in execute
    *args, **kwargs)
psycopg2.DataError: integer out of range


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/srv/observer/src/observer/management/commands/article_update_listener.py", line 23, in action
    ingest_logic.download_article_versions(msid)
  File "/srv/observer/src/observer/ingest_logic.py", line 423, in download_article_versions
    return _download_versions(msid, len(resp["versions"]))
  File "/srv/observer/src/observer/ingest_logic.py", line 418, in _download_versions
    lmap(lambda version: upsert_ajson(consume("articles/%s/versions/%s" % (msid, version))), version_range)
  File "/srv/observer/src/observer/utils.py", line 57, in <lambda>
    lmap = lambda func, *iterable: list(map(func, *iterable))
  File "/srv/observer/src/observer/ingest_logic.py", line 418, in <lambda>
    lmap(lambda version: upsert_ajson(consume("articles/%s/versions/%s" % (msid, version))), version_range)
  File "/srv/observer/src/observer/ingest_logic.py", line 307, in upsert_ajson
    return create_or_update(models.ArticleJSON, article_data, ['msid', 'version'])
  File "/srv/observer/src/observer/utils.py", line 195, in create_or_update
    inst.save()
  File "/srv/observer/venv/lib/python3.5/site-packages/django/db/models/base.py", line 806, in save
    force_update=force_update, update_fields=update_fields)
  File "/srv/observer/venv/lib/python3.5/site-packages/django/db/models/base.py", line 836, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/srv/observer/venv/lib/python3.5/site-packages/django/db/models/base.py", line 922, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/srv/observer/venv/lib/python3.5/site-packages/django/db/models/base.py", line 961, in _do_insert
    using=using, raw=raw)
  File "/srv/observer/venv/lib/python3.5/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
```
The testing id was `3231380270525815893`.

**Comment by lsh-0** [comment]
>   File "/srv/observer/src/observer/ingest_logic.py", line 307, in upsert_ajson
>    return create_or_update(models.ArticleJSON, article_data, ['msid', 'version'])

same model and field that the db migration was created for. has it been applied to this instance yet? because it exists in master: https://github.com/elifesciences/observer/blob/develop/src/observer/migrations/0009_auto_20180109_0335.py

and there are no outstanding PRs that might have unmerged changes in it

**Comment by giorgiosironi** [comment]
Seems to be an `integer` field, and the migration is correctly applied:
```
observerend2end=> \d observer_articlejson
                          Table "public.observer_articlejson"
 Column  |   Type   |                             Modifiers                             
---------+----------+-------------------------------------------------------------------
 id      | integer  | not null default nextval('observer_articlejson_id_seq'::regclass)
 version | smallint | not null
 ajson   | text     | not null
 msid    | integer  | not null
Indexes:
    "observer_articlejson_pkey" PRIMARY KEY, btree (id)
    "observer_articlejson_msid_id_version_70661b7a_uniq" UNIQUE CONSTRAINT, btree (msid, version)
Check constraints:
    "observer_articlejson_msid_d797c2d0_check" CHECK (msid >= 0)
    "observer_articlejson_version_check" CHECK (version >= 0)
observerend2end=> SELECT * FROM django_migrations WHERE name='0009_auto_20180109_0335';
 id |   app    |          name           |            applied            
----+----------+-------------------------+-------------------------------
 22 | observer | 0009_auto_20180109_0335 | 2018-01-09 03:46:21.587108+00
```

**Comment by giorgiosironi** [comment]
`integer` however [only allows `-2147483648 to +2147483647`](https://www.postgresql.org/docs/9.3/static/datatype-numeric.html), we need `bigint` [like in Lax](https://github.com/elifesciences/lax/blob/develop/src/publisher/migrations/0032_auto_20170110_1414.py#L18).

**Comment by lsh-0** [comment]
ah - I didn't realize it was *that* big. I'll get this through today

**Comment by lsh-0** [comment]
not sure what is up with observer end2end, but it knows about 31 article versions and no actual articles. when the article update listener is hit, it should hit lax and run updates on that article. 

```
observerend2end=> select count(*) from articles;
 count 
-------
     0
(1 row)

observerend2end=> select count(*) from observer_articlejson;
 count 
-------
    31
(1 row)
```

inspecting logs

**Comment by lsh-0** [comment]
I've seen this before in lax:
```
[pid: 1503|app: 0|req: 4/4] 10.0.3.151 () {40 vars in 587 bytes} [Fri Jan 12 16:22:52 2018] GET /report/upcoming-articles.rss => generated 644 bytes in 19 msecs (HT
TP/1.0 200) 3 headers in 114 bytes (1 switches on core 1)
2018-01-12 16:44:25,783 (1503/NR-Harvest-Thread) newrelic.core.agent INFO - New Relic Python Agent Shutdown
INFO - New Relic Python Agent Shutdown
Traceback (most recent call last):
  File "/srv/observer/venv/lib/python3.5/site-packages/newrelic-2.78.0.57/newrelic/core/agent.py", line 232, in uwsgi_atexit_callback
    self._atexit_shutdown()
  File "/srv/observer/venv/lib/python3.5/site-packages/newrelic-2.78.0.57/newrelic/core/agent.py", line 670, in _atexit_shutdown
    self.shutdown_agent()
  File "/srv/observer/venv/lib/python3.5/site-packages/newrelic-2.78.0.57/newrelic/core/agent.py", line 682, in shutdown_agent
    self._harvest_thread.join(timeout)
  File "/usr/lib/python3.5/threading.py", line 1051, in join
    raise RuntimeError("cannot join current thread")
RuntimeError: cannot join current thread
DAMN ! worker 1 (pid: 1501) died :( trying respawn ...
Respawned uWSGI worker 1 (new pid: 4760)
DAMN ! worker 2 (pid: 1502) died :( trying respawn ...
Respawned uWSGI worker 2 (new pid: 4761)
DAMN ! worker 3 (pid: 1503) died :( trying respawn ...
Respawned uWSGI worker 3 (new pid: 4762)
DAMN ! worker 4 (pid: 1504) died :( trying respawn ...
Respawned uWSGI worker 4 (new pid: 4763)
DAMN ! worker 2 (pid: 4761) died, killed by signal 9 :( trying respawn ...
Respawned uWSGI worker 2 (new pid: 4829)
DAMN ! worker 4 (pid: 4763) died, killed by signal 9 :( trying respawn ...
Respawned uWSGI worker 4 (new pid: 4830)
```
is one half of a pair of errors I couldn't reconcile. A fix has been merged in for lax, but if the above is being caused by NewRelic then I expect we'll see it again

**Comment by lsh-0** [comment]
looks like article update listener was failing a range check: https://github.com/elifesciences/observer/pull/26

I don't think that log is being monitored either

**Comment by lsh-0** [comment]
observer change has gone through, giving this a kick ...

**Pull request merged**